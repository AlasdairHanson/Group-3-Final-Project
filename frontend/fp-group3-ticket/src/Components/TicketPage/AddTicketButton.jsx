import React, { useState } from "react";
import axios from "axios";
import { Button, Modal, Form } from "react-bootstrap";
import { PlusCircleFill } from "react-bootstrap-icons";
const AddTicketButton = () => {
  const [show, setShow] = useState(false);

  const handleClose = () => setShow(false);
  const handleShow = () => setShow(true);

  const [firstName, setfirstName] = useState(``);
  const [secondName, setsecondName] = useState(``);
  const [cohort, setcohort] = useState(`CloudNative`);
  const [title, settitle] = useState(``);
  const [issue, setissue] = useState(``);
  const [topic, settopic] = useState(`Networking`);
  const [urgency, seturgency] = useState(`Low`);
  const [timestamp, settimestamp] = useState(
    new Date().toLocaleString("en-GB")
  );

  const ticketData = {
    title: title,
    topic: topic,
    content: issue,
    timestamp: timestamp,
    status: "Unresolved",
    author: firstName + " " + secondName,
    trainer: "none",
    priority: urgency,
    cohort: cohort,
  };

  const postData = (e) => {
    const newDate = new Date().toLocaleString("en-GB");
    settimestamp(newDate);
    axios
      .post("http://backend:8081/createTicket", ticketData, {
      headers: {
        'Access-Control-Allow-Origin' : '*'
        }
    })
      .then(function (response) {
        console.log(response);
      })
      .catch(function (error) {
        console.log(error);
      });
    handleClose();
  };

  return (
    <>
      <Button
        variant="primary"
        className="ml-3 qaColButton"
        onClick={handleShow}
      >
        <PlusCircleFill /> New Ticket
      </Button>

      <Modal
        show={show}
        onHide={handleClose}
        dialogClassName="ticketModal"
        scrollable={true}
      >
        <Modal.Header closeButton>
          <Modal.Title>Add Ticket Modal</Modal.Title>
        </Modal.Header>
        <Modal.Body>
          <Form>
            <Form.Group>
              <Form.Label>First Name</Form.Label>
              <Form.Control
                type="text"
                placeholder="First Name"
                onChange={(e) => {
                  setfirstName(e.target.value);
                }}
              />
            </Form.Group>

            <Form.Group>
              <Form.Label>Second Name</Form.Label>
              <Form.Control
                type="text"
                placeholder="Second Name"
                onChange={(e) => {
                  setsecondName(e.target.value);
                }}
              />
            </Form.Group>

            <Form.Group>
              <Form.Label>Cohort</Form.Label>
              <Form.Control
                as="select"
                onChange={(e) => {
                  setcohort(e.target.value);
                }}
              >
                <option>CloudNative</option>
                <option>Software Specialist</option>
                <option>DevOps</option>
                <option>BigData</option>
                <option>Barclays</option>
              </Form.Control>
            </Form.Group>

            <Form.Group>
              <Form.Label>Title summary of the issue</Form.Label>
              <Form.Control
                type="text"
                placeholder="Issue Sumamary"
                onChange={(e) => {
                  settitle(e.target.value);
                }}
              />
            </Form.Group>

            <Form.Group controlId="exampleForm.ControlTextarea1">
              <Form.Label>Description</Form.Label>
              <Form.Control
                as="textarea"
                rows={3}
                onChange={(e) => {
                  setissue(e.target.value);
                }}
              />
            </Form.Group>

            <Form.Group>
              <Form.Label>Topic</Form.Label>
              <Form.Control
                as="select"
                onChange={(e) => {
                  settopic(e.target.value);
                }}
              >
                <option>Networking</option>
                <option>Privacy</option>
                <option>Software</option>
                <option>Hardware</option>
                <option>Keys</option>
              </Form.Control>
            </Form.Group>

            <Form.Group>
              <Form.Label>Urgency</Form.Label>
              <Form.Control
                as="select"
                onChange={(e) => {
                  seturgency(e.target.value);
                }}
              >
                <option>Low</option>
                <option>Medium</option>
                <option>High</option>
                <option>Critical</option>
              </Form.Control>
            </Form.Group>
          </Form>

          <Button
            onClick={(e) => postData(e)}
            variant="primary"
            size="lg"
            block
            className="mb-4"
          >
            Create Ticket
          </Button>
        </Modal.Body>
      </Modal>
    </>
  );
};
export default AddTicketButton;
