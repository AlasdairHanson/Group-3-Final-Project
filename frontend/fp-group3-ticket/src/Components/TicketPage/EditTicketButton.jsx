import React, { useState } from "react";
import axios from "axios";
import { Button, Modal, Form } from "react-bootstrap";
import { PencilFill } from "react-bootstrap-icons";
const EditTicketButton = ({
  id,
  oldTitle,
  oldAuthor,
  oldStatus,
  oldTrainer,
  oldUrgency,
  oldCohort,
  oldDesc,
  oldTopic,
}) => {
  const [show, setShow] = useState(false);

  const handleClose = () => setShow(false);
  const handleShow = () => setShow(true);

  const [cohort, setcohort] = useState(oldCohort);
  const [trainer, settrainer] = useState(oldTrainer);
  const [title, settitle] = useState(oldTitle);
  const [status, setStatus] = useState(oldStatus);
  const [issue, setissue] = useState(oldDesc);
  const [topic, settopic] = useState(oldTopic);
  const [urgency, seturgency] = useState(oldUrgency);
  const [timestamp, settimestamp] = useState(
    new Date().toLocaleString("en-GB")
  );

  const ticketData = {
    title: title,
    topic: topic,
    content: issue,
    status: status,
    trainer: trainer,
    priority: urgency,
    cohort: cohort,
    timestamp: timestamp,
    author: oldAuthor,
  };

  const updateData = (e) => {
    settimestamp(new Date().toLocaleString("en-GB"));
    axios
      .put("http://backend:8081/updateTicket/" + id, ticketData, {
      headers: {
        'Access-Control-Allow-Origin' : '*'
        }
    })
      .then(function (response) {
        console.log(response);
      })
      .catch(function (error) {
        console.log(ticketData);
        console.log(error);
      });
    handleClose();
  };

  return (
    <>
      <Button variant="link" onClick={handleShow}>
        <PencilFill />
      </Button>

      <Modal
        show={show}
        onHide={handleClose}
        dialogClassName="ticketModal"
        scrollable={true}
      >
        <Modal.Header closeButton>
          <Modal.Title>Edit Ticket Modal</Modal.Title>
        </Modal.Header>
        <Modal.Body>
          <Form>
            <Form.Group>
              <Form.Label>Trainer for Ticket</Form.Label>
              <Form.Control
                type="text"
                defaultValue={oldTrainer}
                placeholder="Trainer Name"
                onChange={(e) => {
                  settrainer(e.target.value);
                }}
              />
            </Form.Group>

            <Form.Group>
              <Form.Label>Status</Form.Label>
              <Form.Control
                as="select"
                defaultValue={oldStatus}
                onChange={(e) => {
                  setStatus(e.target.value);
                }}
              >
                <option>Unresolved</option>
                <option>Pending</option>
                <option>Resolved</option>
              </Form.Control>
            </Form.Group>

            <Form.Group>
              <Form.Label>Cohort</Form.Label>
              <Form.Control
                as="select"
                defaultValue={oldCohort}
                onChange={(e) => {
                  setcohort(e.target.value);
                }}
              >
                <option>CloudNative</option>
                <option>Software Specialist</option>
                <option>DevOps</option>
                <option>BigData</option>
                <option>Barclays</option>
              </Form.Control>
            </Form.Group>

            <Form.Group>
              <Form.Label>Title summary of the issue</Form.Label>
              <Form.Control
                type="text"
                placeholder="Issue Summary"
                defaultValue={oldTitle}
                onChange={(e) => {
                  settitle(e.target.value);
                }}
              />
            </Form.Group>

            <Form.Group controlId="exampleForm.ControlTextarea1">
              <Form.Label>Description</Form.Label>
              <Form.Control
                as="textarea"
                rows={3}
                defaultValue={oldDesc}
                onChange={(e) => {
                  setissue(e.target.value);
                }}
              />
            </Form.Group>

            <Form.Group>
              <Form.Label>Topic</Form.Label>
              <Form.Control
                as="select"
                defaultValue={oldTopic}
                onChange={(e) => {
                  settopic(e.target.value);
                }}
              >
                <option>Networking</option>
                <option>Privacy</option>
                <option>Software</option>
                <option>Hardware</option>
                <option>Keys</option>
              </Form.Control>
            </Form.Group>

            <Form.Group>
              <Form.Label>Urgency</Form.Label>
              <Form.Control
                as="select"
                defaultValue={oldUrgency}
                onChange={(e) => {
                  seturgency(e.target.value);
                }}
              >
                <option>Low</option>
                <option>Medium</option>
                <option>High</option>
                <option>Critical</option>
              </Form.Control>
            </Form.Group>
          </Form>
          <Button
            onClick={(e) => updateData(e)}
            variant="primary"
            size="lg"
            block
            className="mb-4"
          >
            Update Ticket
          </Button>
        </Modal.Body>
      </Modal>
    </>
  );
};
export default EditTicketButton;
