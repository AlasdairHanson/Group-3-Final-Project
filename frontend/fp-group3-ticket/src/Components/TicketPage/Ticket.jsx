import axios from "axios";
import React from "react";
import Avatar from "react-avatar";
import { Col, Row, Button, Card, Accordion } from "react-bootstrap";
import * as Icon from "react-bootstrap-icons";
import EditTicketButton from "./EditTicketButton";
const Ticket = ({
  id,
  acc_id,
  title,
  topic,
  desc,
  time,
  status,
  trainee,
  trainer,
  priority,
  cohort,
}) => {
  const deleteTicket = (e) => {
    console.log("delete data");
    axios.delete("http://backend:8081/deleteTicket/" + id, {
      headers: {
        'Access-Control-Allow-Origin' : '*'
        }
    }).then(
      (response) => {
        console.log(response);
      },
      (error) => {
        console.log(error);
      }
    );
  };

  return (
    <Card className="ticket">
      <Card.Header>
        <Row>
          <Accordion.Toggle as={Col} eventKey={acc_id} xs={4}>
            <p className="mutedText">Title:</p>
            <h5 id="tickTitle" className="ticketText">
              {title}
            </h5>
          </Accordion.Toggle>
          <Col xs={2}>
            <div id="tickTopic" className="textContainer blue my-1">
              {topic}
            </div>
          </Col>
          <Col xs={2}>
            <p className="mutedText">Cohort:</p>
            <h6 id="tickCohort" className="ticketText">
              {cohort}
            </h6>
          </Col>
          <Col xs={4} className="icons my-1">
            <EditTicketButton
              id={id}
              oldTitle={title}
              oldAuthor={trainee}
              oldStatus={status}
              oldTrainer={trainer}
              oldUrgency={priority}
              oldCohort={cohort}
              oldDesc={desc}
              oldTopic={topic}
            />
            <Button variant="link" onClick={(e) => deleteTicket(e)}>
              <Icon.TrashFill />
            </Button>
            <div className={"textContainer " + status}>{status}</div>
          </Col>
        </Row>
      </Card.Header>
      <Accordion.Collapse eventKey={acc_id}>
        <Card.Body>
          <Row>
            <Col xs={8}> {desc}</Col>
          </Row>
        </Card.Body>
      </Accordion.Collapse>
      <Card.Footer>
        <Row>
          <Col xs={4}>
            <div className="priorityBar">
              <div className={priority}>
                <p id="tickPriority" className="priorityText">
                  {priority}
                </p>
              </div>
            </div>
          </Col>
          <Col xs={3}>
            <Row>
              <Col xs={2}>
                <Avatar
                  name={trainee}
                  round={true}
                  size={40}
                  textSizeRatio={2.7}
                />
              </Col>
              <Col>
                <p className="mutedText">Author:</p>
                <h6 id="tickTrainee" className="ticketText">
                  {trainee}
                </h6>
              </Col>
            </Row>
          </Col>

          <Col xs={3}>
            <Row>
              <Col xs={2}>
                <Avatar
                  name={trainer}
                  round={true}
                  size={40}
                  textSizeRatio={2.7}
                />
              </Col>
              <Col>
                <p className="mutedText">Trainer:</p>
                <h6 id="tickTrainer" className="ticketText">
                  {trainer}
                </h6>
              </Col>
            </Row>
          </Col>
          <Col xs={2}>
            <p className="mutedText">Timestamp:</p>
            <h6 id="tickTime" className="ticketText">
              {time}
            </h6>
          </Col>
        </Row>
      </Card.Footer>
    </Card>
  );
};
export default Ticket;
