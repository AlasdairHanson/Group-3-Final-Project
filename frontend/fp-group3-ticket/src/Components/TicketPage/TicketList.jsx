import React from "react";
import { Accordion } from "react-bootstrap";
import Ticket from "./Ticket";
const TicketList = ({ data, error, isLoaded }) => {
  if (!isLoaded) {
    return <p> Please wait.... we are loading your information</p>;
  } else {
    return (
      <>
        <Accordion defaultActiveKey="0">
          {data.map((ticket) => (
            <Ticket
              key={ticket.id}
              obj={ticket}
              id={ticket.id}
              acc_id={ticket.id}
              title={ticket.title}
              topic={ticket.topic}
              desc={ticket.content}
              time={ticket.time}
              status={ticket.status}
              trainee={ticket.author}
              trainer={ticket.trainer}
              priority={ticket.priority}
              cohort={ticket.cohort}
            />
          ))}
        </Accordion>
      </>
    );
  }
};
export default TicketList;
