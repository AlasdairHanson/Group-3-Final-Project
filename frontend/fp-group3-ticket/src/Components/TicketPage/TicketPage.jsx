import React, { useEffect, useState } from "react";
import { Row, Col, Container } from "react-bootstrap";
import axios from "axios";
import TicketList from "./TicketList";
import Sidebar from "../SideBar/Sidebar";

const TicketPage = () => {
  const [error, setError] = useState(`false`);
  const [isLoaded, setLoaded] = useState(false);
  const [ticketData, setTicketData] = useState([]);
  const [isUpdate, setIsUpdate] = useState(true);
  const [dataSort, setDataSort] = useState(`/getTicket`);

  const dataSortSetup = (field) => {
    setDataSort("/getTicket" + field);
  };

  useEffect(() => {
    setIsUpdate(false);
    axios.get("http://backend:8081" + dataSort, 
    {
      headers: {
        'Access-Control-Allow-Origin' : '*'
        }
    }
)
      .then(
      (data) => {
        setLoaded(true);
        setError(false);
        setTicketData(data.data);
        console.log(ticketData);
      },
      (error) => {
        setLoaded(true);
        setError(error);
      }
    );
  }, [dataSort]);

  return (
    <Container fluid={true}>
      <Row className="noMargin">
        <Col xs={2} className="mt-3 pt-3 pl-0" sticky="top">
          <Sidebar dataSortSetup={dataSortSetup} />
        </Col>
        <Col xs={10} className="pt-3 mt-3 pr-0">
          <TicketList data={ticketData} error={error} isLoaded={isLoaded} />
        </Col>
      </Row>
    </Container>
  );
};
export default TicketPage;
