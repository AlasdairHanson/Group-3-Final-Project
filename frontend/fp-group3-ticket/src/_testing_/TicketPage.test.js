import { render, screen } from '@testing-library/react';
import TicketPage from '../Components/TicketPage/TicketPage';
import '@testing-library/jest-dom';
import {create} from 'react-test-renderer';

jest.mock(`../Components/TicketPage/TicketList`, () => {
  return function TicketListMock(props){
    return(<p>Mocked TicketList</p>)
  }
});

jest.mock(`../Components/SideBar/Sidebar`, () => {
    return function SidebarMock(props){
      return(<p>Mocked Sidebar</p>)
    }
  });

describe(`Snapshot test of TicketPage`, () => {

  test(` TicketPage should match the snapshot every time it is rendered`, () => {
    const instance = create(<TicketPage/>).toJSON(); 
    expect(instance).toMatchSnapshot();
  });
});