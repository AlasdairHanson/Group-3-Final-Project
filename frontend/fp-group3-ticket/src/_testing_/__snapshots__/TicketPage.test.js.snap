// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test of TicketPage  TicketPage should match the snapshot every time it is rendered 1`] = `
<div
  className="container-fluid"
>
  <div
    className="noMargin row"
  >
    <div
      className="mt-3 pt-3 pl-0 col-2"
      sticky="top"
    >
      <p>
        Mocked Sidebar
      </p>
    </div>
    <div
      className="pt-3 mt-3 pr-0 col-10"
    >
      <p>
        Mocked TicketList
      </p>
    </div>
  </div>
</div>
`;
